From: Niko Tyni <ntyni@debian.org>
Subject: Module::Build::Compat makefiles now support 'distclean'
Closes: 527993

[rt.perl.org #46338]

Passthrough makefiles generated by Module::Build::Compat
now remove the makefile itself in the 'distclean' target.
    
Adapted from upstream svn r12788 and r12789.

---
 lib/Module/Build/Compat.pm  |    5 ++++-
 lib/Module/Build/t/compat.t |   13 +++++++++----
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/lib/Module/Build/Compat.pm b/lib/Module/Build/Compat.pm
index d1bc725..e7819a9 100644
--- a/lib/Module/Build/Compat.pm
+++ b/lib/Module/Build/Compat.pm
@@ -228,13 +228,16 @@ all : force_do_it
 realclean : force_do_it
 	$perl $Build realclean
 	$perl -e unlink -e shift $args{makefile}
+distclean : force_do_it
+	$perl $Build distclean
+	$perl -e unlink -e shift $args{makefile}
 
 force_do_it :
 	@ $noop
 EOF
 
   foreach my $action ($class->known_actions) {
-    next if $action =~ /^(all|realclean|force_do_it)$/;  # Don't double-define
+    next if $action =~ /^(all|distclean|realclean|force_do_it)$/;  # Don't double-define
     $maketext .= <<"EOF";
 $action : force_do_it
 	$perl $Build $action
diff --git a/lib/Module/Build/t/compat.t b/lib/Module/Build/t/compat.t
index 8eb654a..7bd363c 100644
--- a/lib/Module/Build/t/compat.t
+++ b/lib/Module/Build/t/compat.t
@@ -13,7 +13,7 @@ local  @ENV{@makefile_keys};
 delete @ENV{@makefile_keys};
 
 my @makefile_types = qw(small passthrough traditional);
-my $tests_per_type = 14;
+my $tests_per_type = 17;
 if ( $Config{make} && find_in_path($Config{make}) ) {
     plan tests => 38 + @makefile_types*$tests_per_type*2;
 } else {
@@ -249,8 +249,11 @@ sub test_makefile_types {
     ok $success, "make realclean ran without error";
 
     # Try again with some Makefile.PL arguments
-    test_makefile_creation($mb, [], 'INSTALLDIRS=vendor', 1);
+    test_makefile_creation($mb, [], 'INSTALLDIRS=vendor', 'realclean');
     
+    # Try again using distclean
+    test_makefile_creation($mb, [], '', 'distclean');
+
     1 while unlink 'Makefile.PL';
     ok ! -e 'Makefile.PL', "cleaned up Makefile";
   }
@@ -272,10 +275,12 @@ sub test_makefile_creation {
   ok -e 'Makefile', "Makefile exists";
   
   if ($cleanup) {
+    # default to 'realclean' unless we recognize the clean method
+    $cleanup = 'realclean' unless $cleanup =~ /^(dist|real)clean$/;
     $output = stdout_of( sub {
-      $build->do_system(@make, 'realclean');
+      $build->do_system(@make, $cleanup);
     });
-    ok ! -e 'Makefile', "Makefile cleaned up";
+    ok ! -e 'Makefile', "Makefile cleaned up with $cleanup";
   }
   else {
     pass '(skipping cleanup)'; # keep test count constant
-- 
tg: (71b2123..) fixes/module-build-compat (depends on: upstream)
